139-pipeline-identity-tokens/proposal.md

```diff
@@ -0,0 +1,155 @@
+- Concourse exposes the public part of the key as a JWKS ([RFC 7517](https://datatracker.ietf.org/doc/html/rfc7517)) under a publicly accessible path (for example: https://myconcourse.example.com/keys)
+- Concourse offers a minimal OIDC Discovery Endpoint ([RFC8418](https://datatracker.ietf.org/doc/html/rfc8414)) that basically just points to the JWKS-URL
+- There is a built-in var source that pipelines can use to get a signed JWT with the following contents:
+```
+{
+    "iss": "https://myconcourse.example.com",
+    "exp": "expiration-time",
+    "iat": "time-when-token-was-issued",
+    "jti": "nonce",
+    "aud": [<configurable via var-source-config>],
+    "sub": "team/pipeline-name",
+    "team": "team-name",
+    "pipeline": "pipeline-name",
+    ...<whatever else might be relevant>...
+}
+```
+- That JWT is signed with the signature key created in the beginning
+- The jobs/tasks of the pipeline use the token to do whatever they like with it
+
+In the pipeline it would then look like this:
+
+```
+var_sources:
+- name: idtoken
+  type: idtoken
+  config:
+    // the aud claim says what this token is intended for. It must therefore be configurable.
+    aud: ["sts.amazonaws.com"]
+    ttl: 1h
+
+jobs:
+- name: print-creds
+  plan:
+  - task: print
+    config:
+      platform: linux
+      image_resource:
+        type: registry-image
+        source: {repository: ubuntu}
+      run:
+        path: bash
+        args:
+        - -c
+        - |
+          echo token: ((idtoken:token))
+          // Send this token as part of an API-Request to an external service
+```
+
+## Phase 2
+Concourse could periodically rotate the signing key it uses. The new key will then also be published in the JWKS. The previous key MUST also remain published for some time, in case there are still unexpired tokens out there that were signed with it.
+
+## Phase 3
+To make sure tokens are as short-lived as possible we could enable online-verification of tokens. Concourse could offer a Token-Introspection-Endpoint ([RFC7662](https://datatracker.ietf.org/doc/html/rfc7662)) where external services can send tokens to for verification.
+That endpoint could reject any token for a pipeline which is currently not running at all.
+
+# Open Questions
+
+1. How to call the var_source. "idtoken"?
+2. How to call the "field" for the token in the var_source. "token"?
+3. What kind of keypair is used by default? RSA-Keys with 4096 bits? We could generate multiple types of keypairs and allow pipelines to choose one via the var-source-config.
+4. What exactly to put into the sub-claim? The easiest would be "team/pipeline". But what about the job-name or instance-vars? Unfortunately instance-vars and job-names are currently not available to var-sources.
+5. How long should the default ttl be?
+6. How often to rotate the signing key?
+7. Can and should we include more specific information into the token (job-name/id, task, infos about the worker)?
+
```

Comment:
```suggestion
8. How do pipeline identity tokens work with resources?

```

I'm raising this as a question because it wasn't mentioned at all. Since the identity token changes (TTL changes), referencing it in a resource would constantly reset the version history of the resource. This may be fine depending on how the resource behaves.

I don't want identity tokens to change any of the current behaviour of resources, but I do want us to think through how they would work for resources.

I think the `sub` claim should also be different too for resources: `team/pipeline/resource`.

Reply:
Honestly I have no idea how to handle this.
I am open for any suggestions.